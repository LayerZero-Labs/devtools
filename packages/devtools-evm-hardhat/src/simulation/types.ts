import { AnvilOptions } from '@layerzerolabs/devtools-evm'

export interface SimulationUserConfig {
    /**
     * Simulation works by creating local forks of networks specified
     * in hardhat config. These networks are containerized and not accessible
     * from the developer machine - instead, a simple proxy server
     * listening on this port proxies requests to the individual networks.
     *
     * For example, a hardhat configuration with networks `fuji` and `mainnet`
     * will result in two forks being created. These networks will then be accessible
     * on `http://localhost:<port>/fuji` and `http://localhost:<port>/mainnet`
     *
     * @default 8545
     */
    port?: number

    /**
     * Simulation task stores its artifacts on the local filesystem.
     *
     * To customize the path, set this property to a relative or absolute path
     * (relative path will be resolved against the root path of your hardhat project)
     *
     * @default .layerzero
     */
    directory?: string

    /**
     * Simulation will by default overwrite the network accounts by values
     * provided in Anvil config (or defaults if not present).
     *
     * The mnemonic will default to "test test test test test test test test test test test junk"
     * and the number of funded accounts to 10.
     *
     * @default true
     */
    overwriteAccounts?: boolean

    /**
     * Anvil overrides for the underlying EVM nodes.
     */
    anvil?: SimulationAnvilUserConfig
}

/**
 * Resolved simulation config
 */
export interface SimulationConfig {
    port: number
    directory: string
    overwriteAccounts: boolean
    anvil: SimulationAnvilConfig
}

/**
 * Resolved simulation config for anvil.
 *
 * This config is created by taking the user config
 * and applying defaults.
 */
export type SimulationAnvilConfig = AnvilOptions & {
    mnemonic: NonNullable<AnvilOptions['mnemonic']>
}

/**
 * User facing simulation config for anvil.
 *
 * This config cannot override several system attributes for anvil
 * (since it would not really make sense seeing that anvil is being run in containers):
 *
 * - host & port: this would break the docker port bindings
 * - state: state is kept within the container and is not accessible from the outside
 * - forkUrl: this is set based on the hardhat config and points to the hardhat network url
 */
export type SimulationAnvilUserConfig = Omit<AnvilOptions, 'host' | 'port' | 'state' | 'forkUrl'>
