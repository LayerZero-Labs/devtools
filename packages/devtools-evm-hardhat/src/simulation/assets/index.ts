// Even though it would be nice to just call this file Dockerfile instead of Dockerfile.conf,
// esbuild (or tsup) have issues with specifying loaders for files without extensions.
//
// And since we already have a file with .conf extension,
// we add the same extension to the Dockerfile to use the same d.ts file and the same loader
export { default as dockerfile } from './Dockerfile.conf'
export { default as nginxConf } from './nginx.conf'
