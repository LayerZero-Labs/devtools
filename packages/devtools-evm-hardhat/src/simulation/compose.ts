import { pipe } from 'fp-ts/lib/function'
import * as RR from 'fp-ts/ReadonlyRecord'
import type { ComposeSpec, ComposeSpecService } from '@layerzerolabs/devtools'
import { type AnvilOptions, createAnvilCliOptions } from '@layerzerolabs/devtools-evm'
import type { ComposeSpecServices } from '@layerzerolabs/devtools'
import type { SimulationConfig } from './types'

/**
 * Creates a docker compose service specification for an anvil-based EVM node
 *
 * @param {AnvilOptions} anvilOptions
 * @returns {ComposeSpecService}
 */
export const createEvmNodeServiceSpec = (anvilOptions: AnvilOptions): ComposeSpecService => ({
    // This service references a Dockerfile that is copied
    // next to the resulting docker-compose.yaml
    //
    // The source for this Dockerfile is located in src/simulation/assets/Dockerfile.conf
    build: {
        dockerfile: 'Dockerfile',
        target: 'node-evm',
    },
    command: ['anvil', ...createAnvilCliOptions(anvilOptions)],
})

/**
 * Creates a docker compose service specification for an nginx-based proxy service
 * that proxies requests to underlying EVM nodes (or their RPC URLs to be mor precise)
 *
 * @param {number} port
 * @param {ComposeSpecServices} networkServices
 * @returns {ComposeSpecService}
 */
export const createEvmNodeProxyServiceSpec = (
    port: number,
    networkServices: ComposeSpecServices
): ComposeSpecService => ({
    // This service references a Dockerfile that is copied
    // next to the resulting docker-compose.yaml
    //
    // The source for this Dockerfile is located in src/simulation/assets/Dockerfile.conf
    build: {
        dockerfile: 'Dockerfile',
        target: 'proxy-evm',
    },
    // This service will expose its internal 8545 port to a host port
    //
    // The internal 8545 port is hardcoded both here and in the nginx.conf file,
    // the source for which is located in src/simulation/assets/nginx.conf
    ports: [`${port}:8545`],
    depends_on: pipe(
        networkServices,
        // This service will depend on the RPCs to be healthy
        // so we'll take the networkServices object and replace
        // the values with service_healthy condition
        RR.map(() => ({
            condition: 'service_healthy',
        }))
    ),
})

/**
 * Creates a docker compose spec with a set of anvil-based EVM nodes
 * and a single proxy server that proxies requests to these nodes.
 *
 * @param {SimulationConfig} config
 * @param {Record<string, AnvilOptions>} networks
 * @returns {ComposeSpec}
 */
export const createSimulationComposeSpec = (
    config: SimulationConfig,
    networks: Record<string, AnvilOptions>
): ComposeSpec => ({
    version: '3.9',
    services: pipe(
        networks,
        // First we turn the networks into docker compose specs for EVM nodes
        RR.map(createEvmNodeServiceSpec),
        (networkServiceSpecs) =>
            // Then we add the RPC proxy server
            //
            // There is a small edge case here that we can address
            // if it ever comes up: if a network is called 'rpc', this compose file
            // will not work.
            //
            // The fix for this is to prefix all networks with something like network-xxx
            // but we can do that if ever this usecase comes up
            pipe(
                networkServiceSpecs,
                RR.upsertAt('rpc', createEvmNodeProxyServiceSpec(config.port, networkServiceSpecs))
            )
    ),
})
