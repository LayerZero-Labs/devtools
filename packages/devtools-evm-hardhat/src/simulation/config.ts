import type { HardhatConfig, HttpNetworkConfig, NetworkConfig } from 'hardhat/types'
import type { SimulationConfig, SimulationUserConfig } from './types'
import { resolve } from 'path'
import { AnvilOptions } from '@layerzerolabs/devtools-evm'
import { pipe } from 'fp-ts/lib/function'
import * as R from 'fp-ts/Record'

/**
 * Turns `SimulationUserConfig` into `SimulationConfig` by supplying defaults
 *
 * @param {SimulationUserConfig} userConfig
 * @param {HardhatConfig} hardhatConfig
 * @returns {SimulationConfig}
 */
export const resolveSimulationConfig = (
    userConfig: SimulationUserConfig,
    hardhatConfig: HardhatConfig
): SimulationConfig => ({
    port: userConfig.port ?? 8545,
    directory: resolve(hardhatConfig.paths.root, userConfig.directory ?? '.layerzero'),
    overwriteAccounts: userConfig.overwriteAccounts ?? true,
    anvil: {
        // For now we'll hardcode the mnemonic we'll use to seed the accounts on the simulation networks
        mnemonic: 'test test test test test test test test test test test junk',
        ...userConfig.anvil,
        // The host and port need to always point to 0.0.0.0:8545
        // since anvil runs in the container that exposes this port on 0.0.0.0
        host: '0.0.0.0',
        port: 8545,
    },
})

/**
 * Takes a portion of the hardhat networks config and turns the values into `AnvilOptions`
 * to be used by a simulation network container.
 *
 * This is used in the simulation generation phase where the compose files
 * for networks defined in hardhat config are being generated.
 *
 * The `Record<string, NetworkConfig>` is used instead of `NetworksConfig` type from hardhat
 * to avoid any type issues with special keys like `localhost` or `hardhat`.
 *
 * @param {SimulationConfig} config
 * @param {Record<string, NetworkConfig>} networksConfig
 * @returns {Record<string, AnvilOptions>} Object with the same keys as the networks config, with values mapped to `AnvilOptions`
 */
export const getAnvilOptionsFromHardhatNetworks = (
    config: SimulationConfig,
    networksConfig: Record<string, NetworkConfig>
): Record<string, AnvilOptions> =>
    pipe(
        networksConfig,
        // We want to drop all the networks that don't have URLs
        R.filter(isHttpNetworkConfig),
        // And map the network configs into AnvilOptions
        R.map(
            (networkConfig: HttpNetworkConfig): AnvilOptions => ({
                ...config.anvil,
                forkUrl: networkConfig.url,
            })
        )
    )

/**
 * Returns with overrides for hardhat network configuration.
 *
 * This, when merged with a hardhat config, will redirect the network calls to the local EVM nodes.
 *
 * @param {SimulationConfig} config
 * @param {Record<string, NetworkConfig>} networksConfig
 * @returns {Record<string, NetworkConfig>}
 */
export const getHardhatNetworkOverrides = (
    config: SimulationConfig,
    networksConfig: Record<string, NetworkConfig>
): Record<string, NetworkConfig> =>
    pipe(
        networksConfig,
        // We want to drop all the networks that don't have URLs
        R.filter(isHttpNetworkConfig),
        // We'll take the existing network configs and point them to our RPC proxy
        //
        // It's important that these configs are not saved to filesystem as they might contain
        // sensitive data (and forgetting to ignore these files in git could lead to security breaches)
        R.mapWithIndex(
            (networkName, networkConfig): NetworkConfig => ({
                ...networkConfig,
                // We want to redirect this network to the local proxy
                //
                // This is the nginx server listening on the port we configured in the simulation configuration
                url: new URL(networkName, `http://localhost:${config.port}`).toString(),
                accounts: config.overwriteAccounts
                    ? // When overwriting accounts, all the accounts will be switched to the anvil config
                      // (or reasonable defaults if not provided)
                      {
                          mnemonic: config.anvil.mnemonic,
                          // These need to be defaulted to the anvil options
                          // (or the anvil defaults)
                          //
                          // See https://book.getfoundry.sh/reference/cli/anvil for anvil defaults
                          count: config.anvil.count ?? 10,
                          path: config.anvil.derivationPath ?? "m/44'/60'/0'/0/",
                          // These will be hardcoded for now as anvil does not support setting these
                          initialIndex: 0,
                          passphrase: '',
                      }
                    : networkConfig.accounts,
            })
        )
    )

/**
 * Helper utility to pick network configs by their names / object keys.
 *
 * Similar to TypeScript `Pick` helper type, but in runtime.
 *
 * @param {string[]} networks List of networks to pick
 */
export const pickNetworkConfigs = (networks: string[]) =>
    R.filterWithIndex<string, NetworkConfig>((networkName: string) => networks.includes(networkName))

/**
 * Little helper utility that checks whether a network config is not a hardhat network config
 *
 * @param {NetworkConfig} networkConfig
 * @returns {boolean}
 */
const isHttpNetworkConfig = (networkConfig: NetworkConfig): networkConfig is HttpNetworkConfig =>
    'url' in networkConfig && typeof networkConfig.url === 'string'
