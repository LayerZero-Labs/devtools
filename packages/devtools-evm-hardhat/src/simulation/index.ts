export * from './compose'
export * from './config'
export * from './types'
