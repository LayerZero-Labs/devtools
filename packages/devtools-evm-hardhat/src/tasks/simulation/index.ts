export * from './logs'
export * from './start'
export * from './stop'
