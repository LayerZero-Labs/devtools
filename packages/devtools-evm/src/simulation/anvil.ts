import { identity, pipe } from 'fp-ts/lib/function'
import * as A from 'fp-ts/Array'
import * as O from 'fp-ts/Option'

export interface AnvilOptions {
    //
    // Server options
    //

    host?: string
    port?: number

    //
    // Account options
    //

    mnemonic?: string
    count?: number
    derivationPath?: string

    //
    // Forking options
    //

    forkUrl?: string
    forkBlockNumber?: number
    retries?: number
    timeout?: number

    //
    // EVM options
    //

    blockTime?: number

    //
    // State
    //

    state?: string
    stateInterval?: number

    //
    // History
    //

    pruneHistory?: boolean
}

/**
 * Creates a list of CLI arguments for `anvil` command
 * to be used with e.g. `spawn` command
 *
 * @param {AnvilOptions} options
 * @returns {string[]} `anvil` CLI arguments
 */
export const createAnvilCliOptions = ({
    host,
    port,
    mnemonic,
    forkUrl,
    forkBlockNumber,
    retries,
    timeout,
    blockTime,
    count,
    derivationPath,
    state,
    stateInterval,
    pruneHistory,
}: AnvilOptions): string[] =>
    pipe(
        [
            pipe(
                O.fromNullable(host),
                O.map((host) => ['--host', host])
            ),
            pipe(
                O.fromNullable(port),
                O.map((port) => ['--port', String(port)])
            ),
            pipe(
                O.fromNullable(mnemonic),
                O.map((mnemonic) => ['--mnemonic', mnemonic])
            ),
            pipe(
                O.fromNullable(count),
                O.map((count) => ['--count', String(count)])
            ),
            pipe(
                O.fromNullable(derivationPath),
                O.map((derivationPath) => ['--derivation-path', derivationPath])
            ),
            pipe(
                O.fromNullable(forkUrl),
                O.map((forkUrl) => ['--fork-url', forkUrl])
            ),
            pipe(
                O.fromNullable(forkBlockNumber),
                O.map((forkBlockNumber) => ['--fork-block-number', String(forkBlockNumber)])
            ),
            pipe(
                O.fromNullable(retries),
                O.map((retries) => ['--retries', String(retries)])
            ),
            pipe(
                O.fromNullable(timeout),
                O.map((timeout) => ['--timeout', String(timeout)])
            ),
            pipe(
                O.fromNullable(blockTime),
                O.map((blockTime) => ['--block-time', String(blockTime)])
            ),
            pipe(
                O.fromNullable(state),
                O.map((state) => ['--state', state])
            ),
            pipe(
                O.fromNullable(stateInterval),
                O.map((stateInterval) => ['--state-interval', String(stateInterval)])
            ),
            pipe(
                O.fromNullable(pruneHistory),
                O.filter(identity),
                O.map(() => ['--prune-history'])
            ),
        ],
        A.compact,
        A.flatten
    )
