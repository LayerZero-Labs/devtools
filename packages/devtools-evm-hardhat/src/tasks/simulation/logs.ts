import '@/type-extensions'

import { TASK_LZ_TEST_SIMULATION_LOGS, TASK_LZ_TEST_SIMULATION_START } from '@/constants'
import { ActionType } from 'hardhat/types'
import { task } from 'hardhat/config'
import { printLogo } from '@layerzerolabs/io-devtools/swag'
import { LogLevel, createLogger, isFile, printJson, setDefaultLogLevel } from '@layerzerolabs/io-devtools'
import { types } from '@/cli'
import { resolveSimulationConfig } from '@/simulation/config'
import { join } from 'path'
import { spawnSync } from 'child_process'

export interface SimulationLogsTaskArgs {
    logLevel?: LogLevel
}

const action: ActionType<SimulationLogsTaskArgs> = async ({ logLevel = 'info' }, hre) => {
    setDefaultLogLevel(logLevel)

    printLogo()

    const logger = createLogger()

    // Grab the simulation user config from hardhat user config
    const simulationUserConfig = hre.userConfig.layerZero?.experimental?.simulation ?? {}
    logger.verbose(`Using simulation user config:\n${printJson(simulationUserConfig)}`)

    // Resolve the defaults for the simulation config
    const simulationConfig = resolveSimulationConfig(simulationUserConfig, hre.config)
    logger.verbose(`Resolved simulation config:\n${printJson(simulationConfig)}`)

    // Check that the docker compose file exists
    const dockerComposePath = join(simulationConfig.directory, 'docker-compose.yaml')
    if (!isFile(dockerComposePath)) {
        logger.warn(`Could not find simulation docker compose file '${dockerComposePath}'`)
        logger.warn(`Did you run 'npx hardhat ${TASK_LZ_TEST_SIMULATION_START}'?`)

        process.exitCode = 1

        return
    }

    // Spawn docker compose logs command, piping the stdout and stderr to the current shell
    //
    // The error reporting on this part should be improved - we should check that "docker" and "docker compose"
    // are known commands before we go ahead and try executing them
    try {
        logger.verbose(`Spawning docker compose logs command for ${dockerComposePath}`)

        spawnSync('docker', ['compose', '-f', dockerComposePath, 'logs'], {
            stdio: 'inherit',
        })
    } catch (error) {
        logger.error(`Failed to spawn docker compose logs command for ${dockerComposePath}: ${error}`)

        process.exitCode = 1

        return
    }
}

if (process.env.LZ_ENABLE_EXPERIMENTAL_SIMULATION) {
    task(TASK_LZ_TEST_SIMULATION_LOGS, 'Show logs for LayerZero omnichain simulation', action).addParam(
        'logLevel',
        'Logging level. One of: error, warn, info, verbose, debug, silly',
        'info',
        types.logLevel
    )
}
