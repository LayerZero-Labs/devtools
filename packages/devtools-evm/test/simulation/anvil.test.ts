import { AnvilOptions, createAnvilCliOptions } from '@/simulation/anvil'

describe('simulation/anvil', () => {
    describe('createAnvilCliOptions()', () => {
        it('should return an empty array for empty options', () => {
            expect(createAnvilCliOptions({})).toStrictEqual([])
        })

        it('should return an empty array for undefined options', () => {
            // In order to keep this test in sync with the source code, we'll create a type
            // that will require all the properties of AnvilOptions but will make them undefined
            const anvilOptionsWithEverythingUndefined: Record<keyof AnvilOptions, undefined> = {
                host: undefined,
                port: undefined,
                mnemonic: undefined,
                forkUrl: undefined,
                blockTime: undefined,
                count: undefined,
                state: undefined,
                stateInterval: undefined,
            }

            expect(createAnvilCliOptions(anvilOptionsWithEverythingUndefined)).toStrictEqual([])
        })

        const TEST_CASES: AnvilOptions[] = [
            {
                host: '0.0.0.0',
            },
            {
                port: 7777,
            },
            {
                host: 'localhost',
                port: 7777,
            },
            {
                mnemonic: 'oh boi oh boi oh boi oh boi oh boi oh boi',
            },
            {
                mnemonic: 'oh boi oh boi oh boi oh boi oh boi oh boi',
                count: 6,
            },
            {
                forkUrl: 'http://hotmail.com/mainnet',
                blockTime: 1,
            },
        ]

        TEST_CASES.forEach((testCase) => {
            it(`it should work for ${JSON.stringify(testCase)}`, () => {
                expect(createAnvilCliOptions(testCase)).toMatchSnapshot()
            })
        })
    })
})
