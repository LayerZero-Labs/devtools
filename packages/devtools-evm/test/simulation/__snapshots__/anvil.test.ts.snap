// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`simulation/anvil createAnvilCliOptions() it should work for {"forkUrl":"http://hotmail.com/mainnet","blockTime":1} 1`] = `
[
  "--fork-url",
  "http://hotmail.com/mainnet",
  "--block-time",
  "1",
]
`;

exports[`simulation/anvil createAnvilCliOptions() it should work for {"host":"0.0.0.0"} 1`] = `
[
  "--host",
  "0.0.0.0",
]
`;

exports[`simulation/anvil createAnvilCliOptions() it should work for {"host":"localhost","port":7777} 1`] = `
[
  "--host",
  "localhost",
  "--port",
  "7777",
]
`;

exports[`simulation/anvil createAnvilCliOptions() it should work for {"mnemonic":"oh boi oh boi oh boi oh boi oh boi oh boi","count":6} 1`] = `
[
  "--mnemonic",
  "oh boi oh boi oh boi oh boi oh boi oh boi",
  "--count",
  "6",
]
`;

exports[`simulation/anvil createAnvilCliOptions() it should work for {"mnemonic":"oh boi oh boi oh boi oh boi oh boi oh boi"} 1`] = `
[
  "--mnemonic",
  "oh boi oh boi oh boi oh boi oh boi oh boi",
]
`;

exports[`simulation/anvil createAnvilCliOptions() it should work for {"port":7777} 1`] = `
[
  "--port",
  "7777",
]
`;
